### PR TITLE
XenServer: Initial commit for xenserver_guest module unit tests

### PR DESCRIPTION
**Pull request by bvitnik**
- xenserver_guest module: implemented full suite of unit tests.
 - FakeXenAPI lib: greatly extended to support loading and manipulation of
   XenAPI DB sample data thus reducing a number of explicit mocks needed.
 - xenserver_guest_facts module: unit tests refactored and ported to new
   FakeXenAPI.
 - xenserver_guest_powerstate module: unit tests refactored and ported to
   new FakeXenAPI.

##### SUMMARY
Initial commit for xenserver_guest module unit tests and refactoring of unit tests for xenserver_guest_facts and xenserver_guest_powerstate modules.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
xenserver_guest
xenserver_guest_facts
xenserver_guest_powerstate

##### ADDITIONAL INFORMATION
Since xenserver_guest is using a large number of XenAPI methods, mocking all of them individually and keeping track of data manipulation proved to be very hard to do. Mocking any XenAPI method itself is hard to do because of the way method calls are implemented in XenAPI library. They are implemented by dynamically resolving call stack (XenAPI class hierarchy) and instantiating callable objects along the way. Unlike xenserver_guest_facts and xenserver_guest_powerstate modules, for xenserver_guest module I needed a better solution. What I ended up with was to greatly extend FakeXenAPI library to support loading and manipulation of XenAPI DB sample data (kept in .json files). All needed XenAPI methods are implemented directly in FakeXenAPI library where they manipulate in-memory XenAPI DB data, valid during the lifetime of a test function. This way I almost entirely removed mocks. Some mocking is still done but usually mocks are used to wrap functions and keep track of their calls. Unit tests for xenserver_guest_facts and xenserver_guest_powerstate modules were then ported to this new FakeXenAPI library thus greatly reducing the number of mocks, simplifying test functions and producing better coverage.

I understand that some tests now look almost like integration tests but since there is no XenServer simulator and using real XenServer for integration tests would be hard or impossible to do, this is the next best thing.

### CODE REVIEW

**Comment by bvitnik** [comment]
bot_status

File: test/units/modules/cloud/xenserver/test_xenserver_guest_facts_module_main.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
+# -*- coding: utf-8 -*-
```
Comment (felixfontein) [inline test/units/modules/cloud/xenserver/test_xenserver_guest_facts_module_main.py]: Does ansible-test actually find this when PRs for xenserver_guest_facts are found? I know that for example for `module_utils/basic.py`, there's a whole subdirectory containing the tests for it: https://github.com/ansible/ansible/tree/devel/test/units/module_utils/basic

File: test/units/modules/cloud/xenserver/test_xenserver_guest_powerstate.py [hunk 2]
```diff
@@ -1,289 +0,0 @@
-# -*- coding: utf-8 -*-
```
Comment (felixfontein) [inline test/units/modules/cloud/xenserver/test_xenserver_guest_powerstate.py]: Same here.

**Review by felixfontein — COMMENTED** [review]
I'm not sure how exactly the test detection works, so this is potentially a problem. @mattclay?

**Comment by mattclay** [comment]
@felixfontein You can see what `ansible-test` will do with change detection pretty easily:

`ansible-test units --changed-path test/units/modules/cloud/xenserver/common.py --explain -v --python 3.7 --truncate 0`

Specify any single path you want to evaluate, relative to the root of the git tree. You can also use `--changed-from` and pass a filename which lists multiple paths to evaluate.

In the case of the example above the output shows:

```
Detected changes in 1 file(s).
test/units/modules/cloud/xenserver/common.py
Mapping 1 changed file(s) to tests.
test/units/modules/cloud/xenserver/common.py -> units: test/units/modules/cloud/xenserver/
Unit test with Python 3.7
Run command: pytest --boxed -r a -n auto --color yes --junit-xml test/results/junit/python3.7-units.xml -v test/units/modules/cloud/xenserver/test_xenserver_guest_facts.py test/units/modules/cloud/xenserver/test_xenserver_guest_powerstate.py
```

**Comment by felixfontein** [comment]
@mattclay when I run
```
test/runner/ansible-test units --changed-path lib/ansible/modules/cloud/xenserver/xenserver_guest_powerstate.py --explain -v --python 3.7 --truncate 0
```
it does not seem to find all the `test_xenserver_guest_powerstate_module_*.py` tests in `test/units/modules/cloud/xenserver`.

**Comment by mattclay** [comment]
@felixfontein Sorry, I didn't look closely enough at what you were asking earlier. Currently, the code that handles detection of unit tests for modules is very simple compared to what is used for module_utils. For a unit test to match a module it must be in a single file in the tree under `test/units/modules/` named `test_{MODULE_NAME}.py`.

**Comment by bvitnik** [comment]
@mattclay So what are my options here? I can put all the tests in `test_{MODULE_NAME}.py` files, one per module, but whey will be huge and hard to follow. That was the main reason I have divided them into smaller test files. What comes to my mind, but I'm not sure if pytest will be happy with it, is that I could leave all the files as they are but remove `test_` prefix from file names so that pytest does not pick them. Then I can make `test_{MODULE_NAME}.py` files, one per module, and just include all the test functions from other files.

Also, if I have `test_{MODULE_NAME}.py`, will change in module util code trigger tests for modules that depend on it? My thinking is that it should.

**Comment by felixfontein** [comment]
@mattclay would it be possible to extend ansible-test to also check for a directory with the same name as the module, and if it finds that, run all tests in that directory? (I.e. similar to module_utils.)

**Comment by felixfontein** [comment]
I looked at ansible-test a bit, the way it works it is not that easy to implement what I suggested [here](https://github.com/ansible/ansible/pull/56242#issuecomment-505739126).

@bvitnik the idea of having `test_{MODULE_NAME}.py` including `{MODULE_NAME}{suffix}.py` will probably not work well. First, ansible-test assumes that everything has prefix `test_`, so it will screw these files up. Second, when one of these files is changed, it probably will not detect that it has to run the unit tests for `{MODULE_NAME}`.

@bvitnik about your last question "will change in module util code trigger tests for modules that depend on it?": I think changes to module_utils might trigger **all** unit tests. At least that is what I remember from the last time ago I made changes to a module_utils which was specific for a few modules; it triggered a very long test run which felt like it ran everything.

Anyway, for proceeding (I really want to be able to get #57032 merged eventually): how about putting everything into one file, but adding some boilerplate comments inbetween so it is easy to split it up again later?

**Comment by bvitnik** [comment]
@felixfontein If changes in `{MODULE_NAME}{suffix}.py` will not be detected, then it means that changes in files like `common.py`, files in `fixtures` dir, `testcase_` files etc. will also not be detected. That is rather limiting. I will have to experiment with this myself and dive into ansible-test more.

Anyway, since this will require a lot of rework, lets merge #57032. I'll rebase my work on top of it later.

**Comment by felixfontein** [comment]
> @felixfontein If changes in `{MODULE_NAME}{suffix}.py` will not be detected, then it means that changes in files like `common.py`, files in `fixtures` dir, `testcase_` files etc. will also not be detected. That is rather limiting. I will have to experiment with this myself and dive into ansible-test more.

What I wrote is only about modules unit tests. module_utils unit tests are handled differently, and fixtures are also handled differently (I have no idea whether changes to fixtures trigger tests).

See here for some relevant (i.e. the ones I found, no idea if these are all :) ) code pieces:

https://github.com/ansible/ansible/blob/devel/test/runner/lib/classification.py#L183
https://github.com/ansible/ansible/blob/devel/test/runner/lib/classification.py#L190
https://github.com/ansible/ansible/blob/devel/test/runner/lib/target.py#L462-L464

> Anyway, since this will require a lot of rework, lets merge #57032. I'll rebase my work on top of it later.

Ok!

**Comment by felixfontein** [comment]
#57032 is merged, this now needs a rebase. Should not be that complicated, though. :)

**Comment by mattclay** [comment]
@bvitnik @felixfontein Changes to files under `test/units/` which are not themselves unit tests will trigger all unit tests to run in the first parent directory encountered which does contain tests. Changes to module_utils will trigger dependent unit tests based on analysis of import statements.

**Comment by bvitnik** [comment]
@felixfontein @mattclay Hi. I've reorganized unit tests a bit. Test functions are now collected in a single `test_{MODULE_NAME}.py` file per module. To reduce clutter, test cases (data for fixture parameterization) are now located in separate `tescases` module and it's submodules. They are imported by unit tests. I've tested triggers with `ansible-test units --changed-path` and things look OK:

https://termbin.com/kox3 - cat in terminal for colors ;)

Correct tests are triggered for all combinations of changed files. The only disadvantage is that changes in `testcases` will trigger all the tests, even those not using the particular changed test case file. I'm OK with that as long as right tests are among triggered tests.

I'd appreciate any additional comments and suggestions if you have any. Hope this can be merged soon as I have some changes to module code and some new modules but need a good testing ground.

**Comment by felixfontein** [comment]
Hmm, how about moving `FakeXenAPI.py` and `conftest.py` also into a subdirectory, so that changes to them trigger all tests?

**Comment by bvitnik** [comment]
@felixfontein `conftest.py` can't be moved. It's used by pytest to automatically load test fixtures and it needs to be in test dir. root. `FakeXenAPI.py` could be moved but there is no need for it. Changes to it do trigger all the tests. Checkout the link I posted in previous comment. Here is the crucial part:

```
test/units/modules/cloud/xenserver/FakeXenAPI.py -> units: test/units/modules/cloud/xenserver/
```

So every test found in `test/units/modules/cloud/xenserver/` will be run.

**Comment by Akasurde** [comment]
Hi @bvitnik, Thank you very much for your interest in Ansible. This plugin/module is no longer maintained in this repository and has been migrated to https://github.com/ansible-collections/community.general<br><br>If you have further questions please stop by IRC or the mailing list:<br><br>* IRC: #ansible on irc.freenode.net<br>* mailing list: https://groups.google.com/forum/#!forum/ansible-project<br><br>needs_info<br>
